src/utils.hpp

```diff
@@ -125,22 +131,88 @@ template <class Fn> class scope_exit {
     Fn fn_;
 };
 
+template <typename T>
+concept has_to_chars = requires(T v) { std::to_chars(nullptr, nullptr, std::declval<T>()); };
+
+template <typename T>
+concept has_from_chars = requires(T v) { std::from_chars(nullptr, nullptr, std::declval<T>()); };
+
 template <typename StringType, typename T>
 StringType to_string(T value)
-    requires std::is_integral_v<T>
+    requires std::is_integral_v<T> &&
+             (!std::is_same_v<T, bool>) && std::is_same_v<typename StringType::value_type, char> &&
+             std::is_same_v<StringType,
+                 std::basic_string<typename StringType::value_type,
+                     typename StringType::traits_type, typename StringType::allocator_type>>
 {
     // Maximum number of characters required to represent a 64 bit integer as a string
     // 20 bytes for UINT64_MAX or INT64_MIN + null byte
-    static constexpr size_t uint64_max_chars = 21;
+    static constexpr size_t max_chars = 21;
 
-    std::array<char, uint64_max_chars> str{};
+    std::array<char, max_chars> str{};
     auto [ptr, ec] = std::to_chars(str.data(), str.data() + str.size(), value);
-    if (ec == std::errc()) {
+    [[unlikely]] if (ec != std::errc()) {
+        return {};
+    }
+    return {str.data(), ptr};
+}
+
+template <typename StringType, typename T>
+StringType to_string(T value)
+    requires std::is_floating_point_v<T> && std::is_same_v<typename StringType::value_type, char> &&
+             std::is_same_v<StringType,
+                 std::basic_string<typename StringType::value_type,
+                     typename StringType::traits_type, typename StringType::allocator_type>>
+{
+    if constexpr (has_to_chars<double>) {
```

Comment:
you accept T to be floating point, but then only handle the double case here. You could check `has_to_chars(T)` instead.

`max_chars` could be calculated thus:

```cpp
template<typename T>
requires std::is_same_v<T, float> || std::is_same_v<T, double>
// XXX: add long double, though it's tricker, we don't know if it's quad-precision or x87 80-bit long doubles
constexpr std::size_t max_exp_digits = sizeof(T) == 4 ? 2 : 4;

static constexpr std::size_t max_chars =
    std::numeric_limits<T>::digits10 + 1
    + 1 /* sign */ + 1 /* dot */ + 1 /* e */ + 1 /* exp sign */
    +  max_exp_digits<T>;
```